src/chrome/internal/sources/sourceTextRetriever.ts

```diff
@@ -20,47 +20,91 @@ let localize = nls.loadMessageBundle();
-                 * isn't the same.
-                 */
-                logger.warn(`${loadedSource} is associated with several ${printArray('scripts', scripts)} returning the source arbitrarily of the first one`);
-                text = this._scriptSources.getScriptSource(scripts[0]);
-            } else if (loadedSource.contentsLocation === ContentsLocation.PersistentStorage) {
-                // If this is a file, we don't want to cache it, so we return the contents immediately
-                return utils.readFileP(loadedSource.identifier.textRepresentation);
-            } else {
-                // We'll need to figure out what is the right thing to do for SourceScriptRelationship.Unknown
-                throw new LocalizedError('error.sourceText.multipleScriptsNotSupported', localize('error.sourceText.multipleScriptsNotSupported', "Support for getting the text from dynamic sources that have multiple scripts embedded hasn't been implemented yet"));
-            }
+            const sourceRetrievability = this._retrievability(this._scriptSources, loadedSource);
+            text = sourceRetrievability.retrieve();
             this._sourceToText.set(loadedSource, text);
         }
 
         return text;
     }
 
+    public retrievability(source: ILoadedSource): IPossiblyRetrievableText {
+        return this._retrievability(this._scriptSources, source);
+    }
+
     public toString(): string {
         return `Sources text logic\n${printIterable('sources in cache', this._sourceToText.keys())}`;
     }
-}
\ No newline at end of file
+}
+
+export type GetSourceTextRetrievability = (scriptSources: IScriptSourcesRetriever, loadedSource: ILoadedSource) => IPossiblyRetrievableText;
+
+// We use the retrievability instead of the text method to figure out if a loadedSource that we'll send to the client
+// is retrievable, and thus, if it should include a sourceReference in it or not
+export function getSourceTextRetrievability(scriptSources: IScriptSourcesRetriever, loadedSource: ILoadedSource): IPossiblyRetrievableText {
+    const scripts = loadedSource.scriptMapper().scripts;
+    if (loadedSource.sourceScriptRelationship === SourceScriptRelationship.SourceIsSingleScript && scripts.length === 1) {
+        const singleScript = singleElementOfArray(scripts);
+        return new RetrievableText(() => scriptSources.getScriptSource(singleScript));
+    } else if (loadedSource.sourceScriptRelationship === SourceScriptRelationship.SourceIsSingleScript && scripts.length >= 2) {
+        /**
+         * We have two or more scripts associated with this source. At the moment we don't have any further support for this scenario
+         * so we just return the source of the first script. This won't be ideal if for some reason the source of both scripts
+         * isn't the same.
+         */
+        logger.warn(`${loadedSource} is associated with several ${printArray('scripts', scripts)} returning arbitrarily the source of the first one`);
```

Comment:
Maybe log the warning inside the RetrievableText callback so it shows up in the log next to where the text is returned?

Reply:
That makes sense